[WIP] (WiiU) wut toolchain port/Aroma compatibility

ashquarky opened the pull request:
## Description

This PR ports RetroArch to the [wut](https://github.com/devkitPro/wut) toolchain, allowing it to be compiled with current devkitPPC tooling. It also changes the core loading and filesystems such that they are compatible with [Aroma](https://github.com/wiiu-env/Aroma).

There are a few user-visible changes resulting from this, besides the Aroma compatibility:
- Core loading now *only* works on Tiramisu or Aroma CFWs, not classic Mocha or HaxchiFW. Given Tiramisu is a drop-in replacement for these older CFWs this shouldn't pose too much of an issue. Users should make sure they're using an up-to-date version, as the libmocha features RetroArch needs for core loading were only added to Tiramisu after Aroma's release.
- The new core loading method happens to show a WiiU logo and loading music during a core switch. I am undecided if this is an improvement over the black screens of the existing port.
- On Tiramisu, quitting RetroArch after loading a core goes to the Wii U Menu, not HBL. This is due to longstanding incompatibility between ProcUI and HBL. On Aroma, quitting to the menu is the expected behaviour.
- wut has a "socket devoptab" that sits between RetroArch and the underlying socket calls, making them much more POSIX-y and standards-compliant. As a result, all the wiiu-specific networking optimisations are not compatible and have been removed. I have no idea how netplay is affected by the new devoptab. Some of the optimisations could be possible to re-add by overriding `__init_wut_socket`; such changes are not included in this PR.
- libfat is now enabled on a best-effort basis for USB drives (Aroma/Tiramisu) and SD card (Tiramisu only). When using RetroArch on Tiramisu, the SD card is `sd:/`, while on Aroma it's `fs:/vol/external01`. Users switching between Tiramisu and Aroma may have issues with their config file resulting from this difference.
- Symbols are no longer included in the crash handler.

Moving to wut should put RetroArch in a much stronger position for its long-term maintainability, since fixes to the Cafe bindings, stock Cafe devoptab, and networking compatibility will no longer have to be applied to RetroArch specifically, and can instead be adopted through wut updates. In addition, new features should be possible, such as multithreading (wut provides c11 threads) and dynamic core loading. Cores can also access the Cafe API in the standard way for homebrew, rather than being required to copy headers ad-hoc into their own source trees.

## Related Issues

#14429 

## Related Pull Requests

The related PRs are the primary reason this is still a draft.

### Cores
Due to the [removal of `-mwup`](https://github.com/devkitPro/gcc/commit/6da6e2501b375fa1e63cc66f1db86badcc99c85a) in newer devkitPPC tools, many of the core makefiles need to be changed. `-mwup` is directly equivalent to `-D__wiiu__ -DHW_WUP -ffunction-sections -fdata-sections`; though usually the defines can be tweaked a bit. Still need to decide whether `-D__wut__` or similar is worth including.

Every core built for wiiu that currently uses `-mwup` will need to be fixed before merge.

- [x] snes9x (libretro/snes9x#288)
- more cores here? Need to audit which ones are affected

### Toolchain
There's a new docker container drafted up here: https://git.libretro.com/quarky/libretro-build-wiiu/-/tree/wut

I've not filed a PR for this yet (I think I have push access to the real libretro-build-wiiu anyway...). The toolchain container will need to be live for the *same set of nightlies* as this PR's merge - the new tools can't build old RetroArch, and the old tools can't build wut RetroArch. Woohoo!

## Reviewers

@LibretroAdmin 
@gblues (gblues is a co-author of several of the patches submitted here, massive props!)
@Maschell (in case you have comment on the libfat situation)

ashquarky committed:
(WiiU) Core loading support

This commit re-adds core loading support, now compatible with Tiramisu and
Aroma; using a libmocha feature added in those CFWs. Note that this does NOT
work under Haxchi FW, non-payload Mocha versions, etc. RetroArch now requires
either Tiramisu or Aroma on Wii U.

The overall strategy is to make the path of the core to be loaded relative to
the SD card root, pass this off to Mocha, then switch titles to the wrapper
title (usually Health and Safety). In order to pass arguments, sysapp is used,
with a little extra logic (since sysapp passes a blob of arbitrary bytes) to
emulate argc/argv.

# Conflicts:
#	frontend/drivers/platform_wiiu.c

gblues committed:
(WiiU) explicitly shut down HID driver, preventing lockup

== DETAILS
serial debug and some testing revealed that the cause of the
lockup on exit was because of a deadlock in the HID i/o thread.

solution: explicitly shutdown the HID thread during RetroArch
termination process.

vaguerant commented:
> * more cores here? Need to audit which ones are affected

I went through the cores checking which ones still had the legacy `-mwup` flag, here are my results:

# mwup
- [ ] beetle-ngp-libretro
https://github.com/libretro/beetle-ngp-libretro/blob/b5b67e5172daf1c8a5b15c294b16af6ee6ab3a5f/Makefile#L326
- [ ] beetle-pce-fast-libretro:
https://github.com/libretro/beetle-pce-fast-libretro/blob/ba18467f4cba992f9e76d062408ba2022c98bc72/Makefile#L312
- [ ] beetle-pce-libretro:
https://github.com/libretro/beetle-pce-libretro/blob/5aac572a7e609f4b13dfbf66d9763310b2c8a862/Makefile#L269
- [ ] beetle-vba-libretro:
https://github.com/libretro/beetle-vb-libretro/blob/3e845666d7ce235a071eb306e94074f1a72633bf/Makefile#L297
- [ ] beetle-psx-libretro:
https://github.com/libretro/beetle-psx-libretro/blob/17df2b9b93ba20c520c4592deed4531afc37bc7a/Makefile#L306
- [ ] beetle-supergrafx-libretro:
https://github.com/libretro/beetle-supergrafx-libretro/blob/fffbd49fc7b8e3d1977829a5724226cda30149bc/Makefile#L295
- [ ] beetle-wswan-libretro:
https://github.com/libretro/beetle-wswan-libretro/blob/4e73cf52d5223105a8ac175f4332d609364fb5f9/Makefile#L245
- [ ] bk-emulator:
https://github.com/libretro/bk-emulator/blob/cf70ce503ddb65051a58f3c4f61d1e258fc4c84b/Makefile.libretro#L315
- [ ] bsnes-libretro (not currently offered on Wii U anyway):
https://github.com/libretro/bsnes-libretro/blob/9ff5a619a766a6933649c3654026f8ae65dc5a5e/Makefile#L236
- [ ] bsnes-mercury (not currently offered on Wii U anyway):
https://github.com/libretro/bsnes-mercury/blob/fb9a41fe9bc230a07c4506cad3cbf21d3fa635b4/Makefile#L285
- [ ] fbalpha2012_cps3
https://github.com/libretro/fbalpha2012_cps3/blob/5ae6afeda0af01925e42e1f8dc5aca2a825afc51/svn-current/trunk/makefile.libretro#L250
- [ ] FIX94/fixGB (not currently offered on Wii U anyway):
https://github.com/FIX94/fixGB/blob/a83a9c8ff2ec62cbf163e1dbd28abfdb895103e8/libretro/Makefile#L242
- [ ] FIX94/fixNES (not currently offered on Wii U anyway):
https://github.com/FIX94/fixNES/blob/156fcaca9f4cd9c23d423737994c45cfb05d16ca/libretro/Makefile#L247
- [ ] FreeChaF:
https://github.com/libretro/FreeChaF/blob/4d1d5cb83b93728a63f03454e472a23055d9bbfc/Makefile#L209
- [ ] hatari:
https://github.com/libretro/hatari/blob/1ebf0a0488580ef95c0b28f02223b31813c867c5/Makefile.libretro#L161
- [ ] jamsilva/chimerasnes (not currently offered on Wii U anyway):
https://github.com/jamsilva/chimerasnes/blob/0542329f32ab524ce8606c2c859bcd49905a3c98/Makefile#L249
- [ ] libretro-directxbox (not currently offered on Wii U anyway):
https://github.com/libretro/libretro-directxbox/blob/a157f3c1a68cb40dd74d5a664c0e39a245aaa59c/Makefile#L262
- [ ] libretro-meowPC98:
https://github.com/libretro/libretro-meowPC98/blob/cd389b3490d5048873c40edd949a1b0631e0606d/libretro/Makefile.libretro#L183
- [ ] libretro-samples (not currently offered on Wii U anyway):
https://github.com/libretro/libretro-samples/blob/bce193bc1b8c9a3da43b2ead0158a69e28b37ed8/build/make/Makefile#L281
- [ ] mame2000-libretro:
https://github.com/libretro/mame2000-libretro/blob/0208517404e841fce0c094f1a2776a0e1c6c101d/Makefile#L221
- [ ] mame2003_midway:
https://github.com/libretro/mame2003_midway/blob/32e6dc3064e5c7e4259e6474e78774b17ff6abab/Makefile#L161
- [ ] mame2009-libretro (not currently offered on Wii U anyway):
https://github.com/r-type/mame2009-libretro/blob/45048806fbc0fea82709a75a523f2eda945e0a47/Makefile#L269
- [ ] mame2010-libretro (not currently offered on Wii U anyway):
https://github.com/r-type/mame2010-libretro/blob/dd13e65458a41dc2f49d8a3e7cd213ee2b8f0404/Makefile#L292
- [ ] mgba:
https://github.com/libretro/mgba/blob/a69c3434afe8b26cb8f9463077794edfa7d5efad/Makefile.libretro#L343
- [ ] mrboom-libretro:
https://github.com/libretro/mrboom-libretro/blob/1bc0933b71051411404cdc092b14ade17efb2027/Makefile#L239
- [ ] Mu:
https://github.com/libretro/Mu/blob/91a46734dd4083cb8841ff2a6d8874b24203241d/libretroBuildSystem/Makefile.libretro#L357
- [ ] nestopia:
https://github.com/libretro/nestopia/blob/125e2ab6d7f2dd4accc60cb84154adad21d3a6a9/libretro/Makefile#L435
- [ ] nxengine-libretro:
https://github.com/libretro/nxengine-libretro/blob/1b57d2d1d064a78b551e51bb6ec9ea56c185a3e1/Makefile#L315
- [ ] oberon-risc-emu:
https://github.com/libretro/oberon-risc-emu/blob/e82335bba1525b94d928600d758d31d0cfdf5c13/Makefile.libretro#L225
- [ ] picodrive:
https://github.com/libretro/picodrive/blob/b2d43acfbc288038749d8a8fdfbcb0474568e043/Makefile.libretro#L293
- [ ] PokeMini:
https://github.com/libretro/PokeMini/blob/9ba2c2d98bef98794095f3ef50e22f1a3cbc6166/Makefile.libretro#L419
- [ ] quasi88-libretro:
https://github.com/libretro/quasi88-libretro/blob/7980f2484997055a9665837bb4c325d562acb04e/Makefile#L279
- [ ] QuickNES_Core:
https://github.com/libretro/QuickNES_Core/blob/1b88a09f1c386ff9ee46bb371583ae04c5cb5dd0/Makefile#L292
- [ ] REminiscence:
https://github.com/libretro/REminiscence/blob/c2624c7565bbae441835db80f24902fc40f83dd1/Makefile#L230
- [ ] retro8:
https://github.com/libretro/retro8/blob/bc388ec7d217a08265d116aaa74afc0ca3f204f5/Makefile#L188-L189
- [ ] SameBoy:
https://github.com/libretro/SameBoy/blob/09138330990da32362246c7034cf4de2ea0a2a2b/libretro/Makefile#L143
- [ ] scummvm:
https://github.com/libretro/scummvm/blob/e7fe82c4886ea4e7a964f9148cb7d76b90fe1edc/Makefile#L157
- [ ] snes9x2002:
https://github.com/libretro/snes9x2002/blob/540baad622d9833bba7e0696193cb06f5f02f564/Makefile#L137
- [ ] snes9x2005:
https://github.com/libretro/snes9x2005/blob/fd45b0e055bce6cff3acde77414558784e93e7d0/Makefile#L260
- [ ] snes9x2010:
https://github.com/libretro/snes9x2010/blob/e86e54624a7910a64a9a744e3734d4067c48d240/Makefile.libretro#L430
- [ ] stella-emu/stella (not currently offered on Wii U anyway):
https://github.com/stella-emu/stella/blob/d654994c78da5f7ed5d3c2a01913b260989d0aea/src/os/libretro/Makefile#L295
- [ ] stella2014-libretro:
https://github.com/libretro/stella2014-libretro/blob/1351a4fe2ca6b1f3a66c7db0df2ec268ab002d41/Makefile#L165
- [ ] tgbdual-libretro:
https://github.com/libretro/tgbdual-libretro/blob/a6f3018e6a23030afc1873845ee54d4b2d8ec9d3/Makefile#L267
- [ ] timoinutilis/lowres-nx:
https://github.com/timoinutilis/lowres-nx/blob/master/platform/LibRetro/Makefile#L366
- [ ] tyrquake:
https://github.com/libretro/tyrquake/blob/89f3e032757b9b56e8d62a528a4a2472e60631aa/Makefile#L385
- [ ] vba-next:
https://github.com/libretro/vba-next/blob/4191e09e2b0fcf175a15348c1fa8a12bbc6320dd/Makefile.libretro#L391
- [ ] vice-libretro:
https://github.com/libretro/vice-libretro/blob/c34e008a8f4e699c15dda0ce5f6b3bfae0144221/Makefile#L327
- [ ] xmil-libretro:
https://github.com/libretro/xmil-libretro/blob/4cb1e4eaab37321904144d1f1a23b2830268e8df/libretro/Makefile.libretro#L182
- [ ] Zlika/theodore:
https://github.com/Zlika/theodore/blob/7889613edede5ba89de1dfe7c05cf8397cf178ba/Makefile#L452

# no mwup
- [x] 81-libretro
- [x] beetle-lynx-libretro
- [x] blueMSX-libretro
- [x] cannonball
- [x] dosbox-svn
- [x] drhelius/Gearboy
- [x] drhelius/Gearsystem
- [x] ecwolf
- [x] fbalpha2012
- [x] fbalpha2012_cps1
- [x] fbalpha2012_cps2
- [x] fbalpha2012_neogeo
- [x] FBNeo
- [x] fmsx-libretro
- [x] FreeIntv
- [x] frodo-libretro
- [x] fuse-libretro
- [x] gambatte-libretro
- [x] Genesis-Plus-GX
- [x] Genesis-Plus-GX-Wide
- [x] gong
- [x] gpsp
- [x] gw-libretro
- [x] libretro-2048
- [x] libretro-atari800
- [x] libretro-cap32
- [x] libretro-crocods
- [x] libretro-fceumm
- [x] libretro-gme
- [x] libretro-handy
- [x] libretro-lutro
- [x] libretro-o2em
- [x] libretro-pocketcdg
- [x] libretro-prboom
- [x] libretro-uae
- [x] libretro-uzem
- [x] libretro-vecx
- [x] mame2003-libretro
- [x] mame2003-plus-libretro
- [x] nbarkhina/numero (not currently offered on Wii U anyway)
- [x] neocd_libretro
- [x] NP2kai
- [x] prosystem-libretro
- [x] RACE
- [x] snes9x
- [x] SquirrelJME/SquirrelJME
- [x] vitaquake2

Ploggy commented:
Thank You VagueRant, FixGB FixNES are both current WiiU Cores ;) Their on the buildbot anyways :)

Ploggy commented:
I've edited all the makefiles with -mwup and added the new ffunction-sections -fdata-sections -D__wiiu__ -D__wut__ CFLAGS :)
I haven't compiled all the repo to make sure they work with the new changes, but the couple I did compile worked.
I'm not the best with github but I can upload the modified repos if you want?

It'll have to be later tho it's late now and I'm shattered :P

Ploggy commented:
Link to modified Aroma compatible Cores :)
https://easyupload.io/gu7kjs

Ploggy commented:
I finished editing ALL the Cores from the list above with the new changes for Aroma. That took a while lol
Still haven't got a clue if they will all compile..

https://github.com/Ploggy?tab=repositories

ashquarky commented:
Had a discussion with @LibretroAdmin and we're going to try and restructure this so that the old and new buildtypes can coexist to some extent. The goal is still to replace the old toolset eventually but this way we can merge it even with issues remaining and leverage the libretro CI to track progress and make test builds available.

Hopefully with wider testing and more eyes we can work out what's going on with e.g. the very long reported loadtimes.

Ploggy commented:
> Had a discussion with @LibretroAdmin and we're going to try and restructure this so that the old and new buildtypes can coexist to some extent. The goal is still to replace the old toolset eventually but this way we can merge it even with issues remaining and leverage the libretro CI to track progress and make test builds available.
> 
> Hopefully with wider testing and more eyes we can work out what's going on with e.g. the very long reported loadtimes.

Brilliant, saves me manually compiling all the cores :) There were a handful of cores that failed to compile with the new wut changes but they can be tackled as they show up on the CI test builds?

DaveyL2013 commented:
Is this still being worked on?

vaguerant commented:
@DaveyL2013 Probably not right this instant due to other projects, but it's not abandoned.

Ploggy commented:
Hi there :) been a while I just for fun tried a couple of the Aroma Cores and I think I may have stumbled onto something....
I just loaded the FBneo Core with a clean Retroarch.cfg (deleted from SD card prior and let RA create it when loading the Core's rpx from the WiiU Menu) and the rpx loaded super fast and loading a game was normal too????

But how can that be it was soo slow before.. and confirmed by others too?
I pressed the home button to quit and noticed the asset icons were all scrambled as if the paths were incorrect, then I went into Settings>Directory and sure enough the paths seemed to have defaulted to sd:/ instead of fs:/vol/external01. Could that be it could it just be the paths defaulting back to sd:/ and causing the slow load times.

I've only tried this a couple times and maybe it's something else entirely, but could someone else give it a try and confirm?
(I'll put a few Aroma Cores here so you can try it yourself)
This assumes you already have Retroarch setup on your SD Card:
Delete your retroarch.cfg and salamander.cfg if present
Put one of the core rpx in the sd:/wiiu folder
Put the rest in the sd:/retroarch/cores folder
Put sd back in your wiiu, load up Aroma and select the rpx you placed in wiiu folder from the wiiu menu
Retroarch should boot up pretty quickly
Load your rom from load content, select core and it too _should_ load up like normal
but when you press the home button it should look as if the assets are missing and if you look in Settings>Directory the paths will show they have defaulted back to the sd:/

and if you attempt to quit RA it will just stay loading and not go back to the wiiu menu (perhaps if I left it long enough it would have loaded back to the wiiu menu?)

The path defaulting may happen when switching to another core? I seemed to be able to load a few games one after another when using the same core.


[Aroma Cores.zip](https://github.com/user-attachments/files/15518452/Aroma.Cores.zip)

Hopefully this is something trivial and can be fixed easily so this can finally be merged :P

ashquarky commented:
Rebased onto current RetroArch.

ashquarky commented:
Docker build container updated: https://git.libretro.com/quarky/libretro-build-wiiu/-/commit/6c95d67e2ec5ee73f4a4564a2ff81dffb14ef275

Maschell committed:
(WiiU) Implement ProcUI-loop, add home menu (HBM) support

Implements support for the Wii U's HOME button menu (HBM). This fixes various "OS integration" warts
(like powering off the console). Now that the HOME button is used for HBM, change the menu bind to
Start+Select.

Various changes are also required for backgrounding support.

Co-developed-by: Maschell <maschell@gmx.de>
Co-developed-by: Ash Logan <ash@heyquark.com>

ashquarky committed:
(WiiU) Package Salamander as .wuhb along with RA assets

These assets are shared to the cores after core switch. Reading them out of a .wuhb is an Aroma-exclusive feature that makes core loading times dramatically better.

Branding assets provided by @Ploggy, originally by cucholix afaik.

Co-developed-by: Maschell <maschell@gmx.de>

ashquarky commented:
That last change changes how RA is packaged, in exchange for greatly improved load times on Aroma, so I'll quickly go over that.
- Salamander is packed alongside the RetroArch assets into a wuhb bundle, roughly ~80MB in all. This goes into `sd:/wiiu/apps` and is shown on the menu with a generic RetroArch icon. (For those who don't know, Salamander on WiiU just immediately loads the last-used core, it doesn't have any menus or any substantial amount of code at all)
- Cores are still .rpx files, in `sd:/retroarch/cores` - *not* `sd:/wiiu/apps`, which means they no longer appear on the menu. Everyone will now access RetroArch via Salamander, therefore via the wuhb bundle, which due to how wuhb bundles work means that the assets from that bundle are available to every core. This is where the load speed boost comes from - cores load the menu assets from a single file on SD (the wuhb) instead of lots of small files. This works around inefficiencies in the Cafe OS filesystem drivers.

This does mean that users can't "directly" boot a desired core from the menu, instead going via Salamander and their last-used core, but the core switching speed should be fast enough now that this isn't a big problem.

This also sets us up for a potential future where cores are dynamically-linked .rpls, which will practically eliminate load times altogether by putting all the menu code into Salamander.

ashquarky committed:
(WiiU) Add new "Allow Wii U HOME menu" setting

Users who prefer the old legacy behaviour of the Home button opening the RetroArch menu, they can disable the HBM here.

ashquarky commented:
Adding ProcUI/HBM support brings RetroArch in line with official Wii U software as well as all Aroma homebrew. It does create a conflict over the HOME button - it now opens the OS's built in HOME menu overlay rather than the RetroArch menu. To smooth this over, the RetroArch menu is opened alongside the HOME menu so users can access it when they return, and the default "open menu" binding is changed to Start+Select, as it is on the Switch.
For users who prefer the old behaviour, it can still be accessed by disabling Settings -> Input -> Menu -> Allow Wii U HOME menu.

ashquarky commented:
Does anyone know where https://docs.libretro.com/development/retroarch/compilation/wiiu/ are written up? In Git somewhere? I may have to update them if this merges due to the new toolchain

ashquarky committed:
(WiiU) remove libiosuhax (for conversion to subtree)

ashquarky committed:
(wiiu) libmocha: Use relative includes

ashquarky committed:
(wiiu) Port to wut toolchain

This commit is enough to compile RetroArch using wut and more modern
versions of devkitPPC. Among the requisite ports and bugfixes, the wiiu
code has been switched to use wut's h.

There's a handful of headers where RA's are more precise, and these
have been put into the `wiiu/include/` hierarchy mirroring WUT's folder
structure, but with `ra_` prepended to the filename to avoid include
directory priority fighting.

Other changes are related to the header changes:
- update defines to use WUT's new definition
- remove some workarounds/hacks in favor of more precise handling
  in WUT.

Cut features that will be re-added later:
- Current rpx symbol names in the exception handler
- somemopt networking optimisations
- Core loading

gblues committed:
(wiiu) aroma: fix conflict with FTPiiU plug-in

== DETAILS
The HID polling thread was written to be run on a single core, and
basically completely occupies it, not allowing any other threads to
run on it.

It turns out, this is a problem in Aroma, because other things can
be running on your cores. For instance, the FTPiiU server thread runs
on the same core as the HID polling thread, which explains why FTPiiU
stopped working when RetroArch was running.

The other problem, is that if FTPiiU is holding the IPC BufPool mutex
when RA is started, a deadlock will eventually happen because RA will
try to acquire the mutex and the FTPii thread, being suspended, won't
be able to release it.

So!

To fix it:
- add a 500-microsecond sleep to the HID polling thread

ashquarky committed:
(wiiu) libmocha: Backport FSA improvements from modern wut

ashquarky committed:
(wiiu) Pass flags to C++ too and add NDEBUG

C++ files were previously not being compiled with the right flags, which led to slow code and something like 2mb extra (compressed!) binary size. Yikes!

ashquarky committed:
(wiiu) re-enable networking optimisations

wut has somemopt and co now, so we can get a nice speed boost back!

ashquarky committed:
Squashed 'deps/libmocha/' content from commit 89438aa9f3

git-subtree-dir: deps/libmocha
git-subtree-split: 89438aa9f3a9d471ff848422f9dc3db581726127

ashquarky committed:
Merge commit '8ed79217adbe1f954641fee925fb90205b921c57' as 'deps/libmocha'

ashquarky committed:
(wiiu) fix salamander

missed a few things in the initial port

MrHaram commented:
> What are limitations on Wii U side, which prevent us to have fresh MAME core instead of legacy ones?

Would also be interested in knowing
